QC of ETL starting with GDC release 24 clinical tables.

This notebook focuses on the QC of program TARGET data_category clinical

##QC table checklist 


**1. Check schema**

Are all the fields labeled?

Is there a table description?

Do the field labels make sense for all fields?
    
Are the labels correct?

**2. Look at table row number and size**

Do these metrics make sense?

**3. Scroll through table manually**

See if anything stands out - empty columns, etc.

*Note from developer:
There are some columns which are sparsely populated (so they might look empty if you’re just scrolling through the table in the GUI), but there should be at least one non-null entry for every column in every table.*

**4. Number of submitter_id versus BigQuery metadata table**

**5. Number of case_id versus BigQuery metadata table**

**6.Check for any duplicate rows present in the table**

**7. Verify count of table against main program clinical table if available**

**8. Verify submiiter_id count of table against master rel_clinical_data table**

**9. Verify case_id count of table against  master rel_clinical_data table**

##Reference material



*   [NextGenETL](https://github.com/isb-cgc/NextGenETL) GitHub repository
*   [ETL QC SOP draft](https://docs.google.com/document/d/1Wskf3BxJLkMjhIXD62B6_TG9h5KRcSp8jSAGqcCP1lQ/edit)

##Before you begin

You need to load the BigQuery module, authenticate ourselves, create a client variable, and load the necessary libraries.


In [1]:
from google.colab import auth
try:
  auth.authenticate_user()
  print('You have been successfully authenticated!')
except:
  print('You have not been authenticated.')

You have been successfully authenticated!


In [4]:
from google.cloud import bigquery
try:
  project_id = 'isb-project-zero' # Update your_project_number with your project number
  client = bigquery.Client(project=project_id)
  print('BigQuery client successfully initialized')
except:
  print('Failed')

BigQuery client successfully initialized


In [6]:
#Install pypika to build a Query 
!pip install pypika
# Import from PyPika
from pypika import Query, Table, Field, Order

import pandas

     |████████████████████████████████| 61kB 2.0MB/s 
  Created wheel for pypika: filename=PyPika-0.37.7-py2.py3-none-any.whl size=42747 sha256=f326f4f7745b15adde8dbfba0c68295c5435b0bf2c883e7a2f35e9eb37e5194d
  Stored in directory: /root/.cache/pip/wheels/40/b2/20/cf67d3c67186b46241b5069c93da2c9beedbb3f08dba75fffe
Successfully built pypika


## READY TO BEGIN TESTING

##Program TARGET

**Testing Full ID** `isb-project-zero.GDC_Clinical_Data.rel24_clin_TARGET`

[Table location](https://console.cloud.google.com/bigquery?authuser=1&folder=&organizationId=&project=isb-project-zero&p=isb-project-zero&d=GDC_Clinical_Data&t=rel24_clin_TARGET&page=table)

Source : GDC API

Date Created : 	Apr 1, 2020, 7:06:22 PM

Release version : v24


##test 1 - schema verification

**1. Check schema**

Are all the fields labeled?

Is there a table description?

Do the field labels make sense for all fields
    
Are the labels correct

Google documentation column descriptions for [reference](https://cloud.google.com/bigquery/docs/information-schema-tables#column_field_paths_view).

Google documentation table options for [reference](https://cloud.google.com/bigquery/docs/information-schema-tables#options_table).

In [0]:
#return all table information for rel24_clin_TARGET

clin_table = Table('`isb-project-zero`.GDC_Clinical_Data.INFORMATION_SCHEMA.TABLES')
clin_query = Query.from_(clin_table) \
                  .select(' table_catalog, table_schema, table_name, table_type ') \
                  .where(clin_table.table_name=='rel24_clin_TARGET') \
                  
clin_query_clean = str(clin_query).replace('"', "")
clin = client.query(clin_query_clean).to_dataframe()
clin.head()

table_catalog       table_schema         table_name  table_type
0  isb-project-zero  GDC_Clinical_Data  rel24_clin_TARGET  BASE TABLE

In [0]:
#return all table information for rel24_clin_TARGET

clin_table = Table('`isb-project-zero`.GDC_Clinical_Data.INFORMATION_SCHEMA.TABLE_OPTIONS')
clin_query = Query.from_(clin_table) \
                  .select(' table_name, option_name, option_type, option_value ') \
                  .where(clin_table.table_name=='rel24_clin_TARGET') \

clin_query_clean = str(clin_query).replace('"', "")
clin = client.query(clin_query_clean).to_dataframe()
pandas.options.display.max_rows


for i in range(len(clin)):
    print(clin['option_name'][i] + '\n')
    print('\t' + clin['option_value'][i] + '\n')
    print('\t' + clin['option_type'][i] + '\n')

else:

    print('QC of friendly name, table description and labels --- FAILED')

QC of friendly name, table description and labels --- FAILED


In [0]:
#check for empty schemas in dataset rel24_clin_TARGET

clin_table = Table('`isb-project-zero`.GDC_Clinical_Data.INFORMATION_SCHEMA.TABLE_OPTIONS')
clin_query = Query.from_(clin_table) \
                  .select(' table_name, option_name, option_type, option_value ') \
                  .where(clin_table.table_name=='rel24_clin_TARGET') \

clin_query_clean = str(clin_query).replace('"', "")
clin = client.query(clin_query_clean).to_dataframe()
pandas.options.display.max_rows
print("Are there any empty cells in the table schema?")
clin.empty

Are there any empty cells in the table schema?


True

FIELD Descriptions pulled example below


In [0]:
#list of field descriptions for table rel24_clin_TARGET

clin_table = Table('`isb-project-zero`.GDC_Clinical_Data.INFORMATION_SCHEMA.COLUMN_FIELD_PATHS')
clin_query = Query.from_(clin_table) \
                  .select('table_name, column_name, description') \
                  .where(clin_table.table_name=='rel24_clin_TARGET') \

clin_query_clean = str(clin_query).replace('"', "")
clin = client.query(clin_query_clean).to_dataframe()
pandas.options.display.max_rows


for i in range(len(clin)):
  print(clin['table_name'][i] + '\n')
  print('\t' + clin['column_name'][i] + '\n')
  print('\t' + clin['description'][i] + '\n')

rel24_clin_TARGET

	submitter_id

	

rel24_clin_TARGET

	case_id

	

rel24_clin_TARGET

	primary_site

	

rel24_clin_TARGET

	disease_type

	

rel24_clin_TARGET

	demo__demographic_id

	

rel24_clin_TARGET

	demo__gender

	Text designations that identify gender. Gender is described as the assemblage of properties that distinguish people on the basis of their societal roles. [Explanatory Comment 1: Identification of gender is based upon self-report and may come from a form, questionnaire, interview, etc.]

rel24_clin_TARGET

	demo__race

	An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bure

In [7]:
# check for empty schemas in dataset rel24_clin_TARGET

clin_table = Table('`isb-project-zero`.GDC_Clinical_Data.INFORMATION_SCHEMA.COLUMN_FIELD_PATHS')
clin_query = Query.from_(clin_table) \
                  .select('table_name, column_name, description') \
                  .where(clin_table.table_name=='rel23_clin_TARGET') \

clin_query_clean = str(clin_query).replace('"', "")
clin = client.query(clin_query_clean).to_dataframe()

print(clin)

           table_name  ...                                        description
0   rel23_clin_TARGET  ...                                                   
1   rel23_clin_TARGET  ...                                                   
2   rel23_clin_TARGET  ...                                                   
3   rel23_clin_TARGET  ...                                                   
4   rel23_clin_TARGET  ...                                                   
5   rel23_clin_TARGET  ...  Text designations that identify gender. Gender...
6   rel23_clin_TARGET  ...  An arbitrary classification of a taxonomic gro...
7   rel23_clin_TARGET  ...  An individual's self-described social and cult...
8   rel23_clin_TARGET  ...  The survival state of the person registered on...
9   rel23_clin_TARGET  ...  Number of days between the date used for index...
10  rel23_clin_TARGET  ...  Number of days between the date used for index...
11  rel23_clin_TARGET  ...                   The current state o

##test 2 row number verification

**2. Look at table row number and size**

Do these metrics make sense?

### pass

In [0]:
%%bigquery --project isb-project-zero
SELECT COUNT(submitter_id)
FROM `isb-project-zero.GDC_Clinical_Data.rel24_clin_TARGET`

f0_
0  6197

In [0]:
%%bigquery --project isb-project-zero
SELECT COUNT(case_id)
FROM `isb-project-zero.GDC_Clinical_Data.rel23_clin_TARGET`

f0_
0  6197

In [0]:
%%bigquery --project isb-project-zero
SELECT *
FROM `isb-project-zero.GDC_Clinical_Data.rel24_clin_TARGET`

submitter_id  ...                  updated_datetime
0          TARGET-10-PAPDAU  ...  2019-08-21T12:39:49.301216-05:00
1         TARGET-00-RO02361  ...  2019-10-24T08:22:10.208559-05:00
2         TARGET-00-RO02590  ...  2019-10-24T08:22:10.208559-05:00
3         TARGET-00-RO02208  ...  2019-10-24T08:22:10.208559-05:00
4         TARGET-00-RO02356  ...  2019-10-24T08:22:10.208559-05:00
...                     ...  ...                               ...
6192  TARGET-20-SJAML045738  ...  2019-07-31T16:37:20.562547-05:00
6193  TARGET-20-SJAML045739  ...  2019-07-31T16:37:20.562547-05:00
6194  TARGET-20-SJAML045740  ...  2019-07-31T16:37:20.562547-05:00
6195  TARGET-20-SJAML045741  ...  2019-07-31T16:37:20.562547-05:00
6196  TARGET-20-SJAML045742  ...  2019-07-31T16:37:20.562547-05:00

[6197 rows x 35 columns]

##test 3 - manual verification

**3. Scroll through table manually**

See if anything stands out - empty columns, etc.

The BigQuery table search user interface is useful in for this test run. The test tier points to the isb-etl-open. 

ISB-CGC BigQuery table  search [test tier](https://isb-cgc-test.appspot.com/bq_meta_search/).

BigQuery console [isb-project-zero](https://console.cloud.google.com/bigquery?authuser=1&folder=&organizationId=&project=isb-project-zero&p=isb-project-zero&d=GDC_Clinical_Data&t=rel24_clin_TARGET&page=table).

Run a manual check in the console with the steps mentioned in step 1 

Are all the fields labeled?

Is there a table description?

Do the field labels make sense for all fields?
    
Are the labels correct?

*Note from developer:
There are some columns which are sparsely populated (so they might look empty if you’re just scrolling through the table in the GUI), but there should be at least one non-null entry for every column in every table.*


##test 4 - submitter_id file metadata table count verification

**4. Number of submitter_id cases versus BigQuery metadata table**

In [0]:
# clinical submitter_id counts table reuslts below

# Query below will display the number of cases presents in this table.

clin_table = Table('`isb-project-zero.GDC_Clinical_Data.rel24_clin_TARGET`')
clin_query = Query.from_(clin_table) \
                  .select(' DISTINCT submitter_id, count(*) as count') \
                  .groupby('submitter_id')

clin_query_clean = str(clin_query).replace('"', "")
#print(clin_query_clean)
clin = client.query(clin_query_clean).to_dataframe()
print('number of cases from submitter_id = ' + str(len(clin.index)))


number of cases from submitter_id = 6140


In [0]:
# GDC file metadata table submitter_id count for clinical below

%%bigquery --project isb-project-zero

SELECT case_barcode, program_name
FROM `isb-project-zero.GDC_metadata.rel24_caseData`
where program_name = 'TARGET'
and active_file_count != 0
group by case_barcode, program_name

case_barcode program_name
0           TARGET-00-NAAENF       TARGET
1           TARGET-00-NAAENG       TARGET
2           TARGET-00-NAAENH       TARGET
3           TARGET-00-NAAENI       TARGET
4           TARGET-00-NAAENJ       TARGET
...                      ...          ...
6099  TARGET-15-SJMPAL016343       TARGET
6100  TARGET-15-SJMPAL040028       TARGET
6101  TARGET-15-SJMPAL040030       TARGET
6102  TARGET-15-SJMPAL040031       TARGET
6103  TARGET-15-SJMPAL040036       TARGET

[6104 rows x 2 columns]

### no match rel24_clin_TARGET: 6140 and rel24_caseData: 6104

##test 5 - case_gdc_id file metadata table count verification

**5. Number of case_id versus BigQuery metadata table**

In [0]:
# clinical case_id counts table reuslts below

# Query below will display the number of cases presents in this table.

clin_table = Table('`isb-project-zero.GDC_Clinical_Data.rel24_clin_TARGET`')
clin_query = Query.from_(clin_table) \
                  .select(' DISTINCT case_id, count(*) as count') \
                  .groupby('case_id')

clin_query_clean = str(clin_query).replace('"', "")
#print(clin_query_clean)
clin = client.query(clin_query_clean).to_dataframe()
print('number of case from case_id = ' + str(len(clin.index)))


number of case from case_id = 6197


In [0]:
# GDC file metadata table case_gdc_id count for clinical below

%%bigquery --project isb-project-zero
SELECT case_gdc_id, program_name
FROM `isb-project-zero.GDC_metadata.rel24_caseData`
where program_name = 'TARGET'
and active_file_count != 0
group by case_gdc_id, program_name

case_gdc_id program_name
0     ef4b5ddc-60f7-46f9-96dc-ab195a94076d       TARGET
1     f48c43f4-715b-42bd-bacf-ec0e545a8d0d       TARGET
2     78246c3a-6485-4791-9a7b-35e5a9a8e91a       TARGET
3     f811637d-f219-4762-a73d-2553f6400205       TARGET
4     f11fa345-a9df-40ec-a3aa-393654dbcf0a       TARGET
...                                    ...          ...
6109  11823197-0ad7-4951-993d-0b4278d991e7       TARGET
6110  bdd81a45-511a-414b-9beb-aa53be092593       TARGET
6111  3219ef77-f370-48b5-bf41-2f3e36543330       TARGET
6112  5993da0e-ebb5-48c7-914f-7a798b907780       TARGET
6113  538662d9-310a-4de4-bda9-5793162b40fb       TARGET

[6114 rows x 2 columns]

### no match rel24_clin_TARGET:6197 and rel24_caseData:6114

##test 6 - duplication verifcation

**6. Check for any duplicate rows present in the table**



###pass

In [0]:
%%bigquery --project isb-project-zero

SELECT count(submitter_id) as count
FROM `isb-project-zero.GDC_Clinical_Data.rel24_clin_TARGET`
group by submitter_id, case_id, primary_site, disease_type, demo__demographic_id, demo__gender, demo__race, demo__ethnicity, demo__vital_status, demo__days_to_birth, demo__days_to_death, demo__state, demo__created_datetime, demo__updated_datetime, diag__diagnosis_id, diag__primary_diagnosis, diag__year_of_diagnosis, diag__progression_or_recurrence, diag__inss_stage, diag__site_of_resection_or_biopsy, diag__age_at_diagnosis, diag__days_to_last_follow_up, diag__tumor_grade, diag__last_known_disease_status, diag__morphology, diag__classification_of_tumor, diag__cog_renal_stage, diag__tumor_stage, diag__tissue_or_organ_of_origin, diag__state, diag__created_datetime, diag__updated_datetime, state, created_datetime, updated_datetime
order by count desc
LIMIT 10


count
0      1
1      1
2      1
3      1
4      1
5      1
6      1
7      1
8      1
9      1

## test 7 - one to many tables count verifcation

**7. Verify count of table against main program clinical table if available**

In [0]:
#no one to many tables avaialable for program TARGET data_category clinical

##test 8 - submitter_id master clinical data table count verifcation

**8. Verify submiiter_id count of table against master rel_clinical_data table**

In [0]:
# submitter_id count from the program TARGET clinical table

%%bigquery --project isb-project-zero

select distinct submitter_id, count(submitter_id) as count
from `isb-project-zero.GDC_Clinical_Data.rel24_clin_TARGET` 
group by submitter_id
order by count

submitter_id  count
0     TARGET-00-RO02361      1
1     TARGET-00-RO02590      1
2     TARGET-00-RO02208      1
3     TARGET-00-RO02356      1
4      TARGET-40-NAAWDN      1
...                 ...    ...
6135   TARGET-10-PAKRSL      2
6136   TARGET-10-PALKTY      2
6137   TARGET-10-PALNTB      2
6138   TARGET-10-PAMDKS      2
6139   TARGET-10-PALJCF      2

[6140 rows x 2 columns]

In [0]:
# submitter_id count from the master clinical table

%%bigquery --project isb-project-zero

SELECT distinct submitter_id, count(submitter_id) as count
FROM `isb-project-zero.GDC_metadata.rel24_caseData` as caseData, `isb-project-zero.GDC_Clinical_Data.rel24_clinical_data` as clinical
WHERE active_file_count != 0 and program_name = 'TARGET' 
and caseData.case_barcode = clinical.submitter_id
group by submitter_id
order by count

submitter_id  count
0     TARGET-00-NAAENF      1
1     TARGET-00-NAAENG      1
2     TARGET-00-NAAENH      1
3     TARGET-00-NAAENI      1
4     TARGET-00-NAAENJ      1
...                ...    ...
6099  TARGET-10-PALKTY      4
6100  TARGET-10-PALLSD      4
6101  TARGET-10-PALNTB      4
6102  TARGET-10-PALTWS      4
6103  TARGET-10-PAMDKS      4

[6104 rows x 2 columns]

### match

##test 9 - case_id master clinical data table count verifcation

**9. Verify case_id count of table against master rel_clinical_data table**

In [0]:
# case_id count from the program TARGET clinical table

%%bigquery --project isb-project-zero

select distinct case_id, count(case_id) as count
from `isb-project-zero.GDC_Clinical_Data.rel24_clin_TARGET` 
group by case_id
order by count

case_id  count
0     56ef0d3a-0829-4870-82ab-dd7ec22fc24b      1
1     d6421ba5-e779-4574-b0a3-d8ffb8b7f9ab      1
2     9d345bf9-7cd9-47eb-942b-72caeab81d56      1
3     d765cc1a-78ef-49e1-9e71-e16c7bf80611      1
4     741ab440-c16e-42df-965e-31827a45ef55      1
...                                    ...    ...
6192  daa5b54a-30b0-44e2-84b7-d186bb4dac62      1
6193  41fbb10c-0786-4c19-8bdd-0f534c3631fd      1
6194  8c456f6d-43cc-4864-acf8-0313a3119050      1
6195  c5723866-f60a-4d7d-bfbd-9d196caa7d7c      1
6196  e8116a9a-0b3e-4c5d-a895-432806794124      1

[6197 rows x 2 columns]

In [0]:
# case_id count from the master clinical table

%%bigquery --project isb-project-zero

SELECT distinct case_id, count(case_id) as count
FROM `isb-project-zero.GDC_metadata.rel24_caseData` as active, `isb-project-zero.GDC_Clinical_Data.rel24_clinical_data` as clinical
WHERE active_file_count != 0 and program_name = 'TARGET'
AND active.case_gdc_id = clinical.case_id
group by case_id
order by count


case_id  count
0     ef4b5ddc-60f7-46f9-96dc-ab195a94076d      1
1     f48c43f4-715b-42bd-bacf-ec0e545a8d0d      1
2     78246c3a-6485-4791-9a7b-35e5a9a8e91a      1
3     f811637d-f219-4762-a73d-2553f6400205      1
4     f11fa345-a9df-40ec-a3aa-393654dbcf0a      1
...                                    ...    ...
6109  11823197-0ad7-4951-993d-0b4278d991e7      1
6110  bdd81a45-511a-414b-9beb-aa53be092593      1
6111  3219ef77-f370-48b5-bf41-2f3e36543330      1
6112  5993da0e-ebb5-48c7-914f-7a798b907780      1
6113  538662d9-310a-4de4-bda9-5793162b40fb      1

[6114 rows x 2 columns]

### no match rel24_clin_TARGET: 6197 and rel24_fileData_current: 6114